Sometimes, the revision-gateway times out without result.

I wrote two tests and the first run showed an error:

https://github.com/holixon/axon-gateway-extension/actions/runs/2516680372

So I disabled the tests again... Needs further attention!

```
2022-06-17 16:15:38.630  INFO 1764 --- [           main] i.h.a.g.example.XstreamApplicationITest  : Starting XstreamApplicationITest using Java 11.0.5 on fv-az449-931 with PID 1764 (started by runner in /home/runner/work/axon-gateway-extension/axon-gateway-extension/example)
2022-06-17 16:15:38.630  INFO 1764 --- [           main] i.h.a.g.example.XstreamApplicationITest  : No active profile set, falling back to 1 default profile: "default"
2022-06-17 16:15:38.995  INFO 1764 --- [           main] o.s.c.a.ConfigurationClassPostProcessor  : Cannot enhance @Configuration bean definition 'org.axonframework.springboot.autoconfig.EventProcessingAutoConfiguration' since its singleton instance has been created too early. The typical cause is a non-static @Bean method with a BeanDefinitionRegistryPostProcessor return type: Consider declaring such methods as 'static'.
2022-06-17 16:15:39.031  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '(inner bean)#71d99732#31' of type [org.axonframework.spring.config.annotation.SpringContextParameterResolverFactoryBuilder$ClasspathParameterResolverFactoryBean] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.032  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '(inner bean)#71d99732#31' of type [org.axonframework.messaging.annotation.MultiParameterResolverFactory] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.035  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '(inner bean)#2328ec73#31' of type [org.axonframework.spring.config.annotation.SpringBeanDependencyResolverFactory] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.037  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '(inner bean)#762da62b' of type [org.axonframework.spring.config.annotation.SpringBeanParameterResolverFactory] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.041  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '__axon-parameter-resolver-factory' of type [org.axonframework.spring.config.ApplicationContextLookupParameterResolverFactory] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.042  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '__axon-parameter-resolver-factory' of type [org.axonframework.messaging.annotation.MultiParameterResolverFactory] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.047  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '(inner bean)#1fb71788' of type [org.axonframework.spring.config.annotation.SpringHandlerDefinitionBean] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.047  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '(inner bean)#1fb71788' of type [org.axonframework.messaging.annotation.MultiHandlerDefinition] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.049  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '(inner bean)#bb50de6' of type [org.axonframework.spring.config.annotation.SpringHandlerEnhancerDefinitionBean] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.049  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '(inner bean)#bb50de6' of type [org.axonframework.messaging.annotation.MultiHandlerEnhancerDefinition] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.052  INFO 1764 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '__axon-handler-definition' of type [org.axonframework.messaging.annotation.MultiHandlerDefinition] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2022-06-17 16:15:39.129  INFO 1764 --- [           main] q.RevisionAwareQueryGatewayConfiguration : REVISION-QUERY-GATEWAY-001: Using revision-aware query gateway.
2022-06-17 16:15:39.834  INFO 1764 --- [           main] o.a.s.a.XStreamAutoConfiguration         : Initializing an XStream instance since none was found. The auto configuration base packages will be used as wildcards for the XStream security settings.
2022-06-17 16:15:39.942  INFO 1764 --- [           main] o.a.s.util.XStreamSecurityTypeUtility    : Constructing wildcard type for base package [io.holixon.axon.gateway.example].
2022-06-17 16:15:39.949  INFO 1764 --- [           main] o.a.serialization.ChainingConverter      : ContentTypeConverter of type [class org.axonframework.serialization.xml.Dom4JToByteArrayConverter] is ignored. It seems to rely on a class that is not available in the class loader: org/dom4j/Document
2022-06-17 16:15:39.949  INFO 1764 --- [           main] o.a.serialization.ChainingConverter      : ContentTypeConverter of type [class org.axonframework.serialization.xml.InputStreamToDom4jConverter] is ignored. It seems to rely on a class that is not available in the class loader: org/dom4j/Document
2022-06-17 16:15:39.950  INFO 1764 --- [           main] o.a.serialization.ChainingConverter      : ContentTypeConverter of type [class org.axonframework.serialization.xml.XomToStringConverter] is ignored. It seems to rely on a class that is not available in the class loader: nu/xom/Document
2022-06-17 16:15:39.951  INFO 1764 --- [           main] o.a.serialization.ChainingConverter      : ContentTypeConverter of type [class org.axonframework.serialization.xml.InputStreamToXomConverter] is ignored. It seems to rely on a class that is not available in the class loader: nu/xom/ParsingException
2022-06-17 16:15:40.138  INFO 1764 --- [           main] i.h.a.g.example.XstreamApplicationITest  : Started XstreamApplicationITest in 1.539 seconds (JVM running for 8.799)
2022-06-17 16:15:40.146  INFO 1764 --- [eway.example]-0] o.a.e.TrackingEventProcessor             : Worker assigned to segment Segment[0/0] for processing
2022-06-17 16:15:40.158  INFO 1764 --- [eway.example]-0] o.a.e.TrackingEventProcessor             : Using current Thread for last segment worker: TrackingSegmentWorker{processor=io.holixon.axon.gateway.example, segment=Segment[0/0]}
2022-06-17 16:15:40.158  INFO 1764 --- [eway.example]-0] o.a.e.TrackingEventProcessor             : Fetched token: null for segment: Segment[0/0]
2022-06-17 16:15:40.191  INFO 1764 --- [           main] i.h.axon.gateway.example.TestScenario    : Sending create command for 43f7c222-346e-49e6-9126-20455f5de6a2 with metadata 'projectionRevision'->'1'
2022-06-17 16:15:40.199  INFO 1764 --- [eway.example]-0] i.h.a.g.e.ApprovalRequestProjection      : Received 43f7c222-346e-49e6-9126-20455f5de6a2
2022-06-17 16:15:40.220  INFO 1764 --- [           main] o.a.e.TrackingEventProcessor             : Shutdown state set for Processor 'io.holixon.axon.gateway.example'.
2022-06-17 16:15:40.226  INFO 1764 --- [       Thread-3] o.a.e.TrackingEventProcessor             : Processor 'io.holixon.axon.gateway.example' awaiting termination...
2022-06-17 16:15:41.201  INFO 1764 --- [eway.example]-0] o.a.e.TrackingEventProcessor             : Released claim
2022-06-17 16:15:41.201  INFO 1764 --- [eway.example]-0] o.a.e.TrackingEventProcessor             : Worker for segment Segment[0/0] stopped.
Error:  Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.629 s <<< FAILURE! - in io.holixon.axon.gateway.example.XstreamApplicationITest
Error:  runScenario  Time elapsed: 0.053 s  <<< ERROR!
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: Approval request with id 43f7c222-346e-49e6-9126-20455f5de6a2 not found.
	at io.holixon.axon.gateway.example.XstreamApplicationITest.runScenario(XstreamApplicationITest.kt:19)
Caused by: java.lang.IllegalArgumentException: Approval request with id 43f7c222-346e-49e6-9126-20455f5de6a2 not found.
	at io.holixon.axon.gateway.example.XstreamApplicationITest.runScenario(XstreamApplicationITest.kt:19)
[INFO] 
[INFO] Results:
[INFO] 
Error:  Errors: 
Error:    XstreamApplicationITest.runScenario:19 » Completion java.lang.IllegalArgumentE...
[INFO] 
Error:  Tests run: 2, Failures: 0, Errors: 1, Skipped: 0
[INFO] 
```

--- a/example/pom.xml
+++ b/example/pom.xml
@@ -69,6 +69,10 @@
       <artifactId>spring-test</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.awaitility</groupId>
+      <artifactId>awaitility-kotlin</artifactId>
+    </dependency>
   </dependencies>
 
   <build>

--- a/example/src/main/kotlin/io/holixon/axon/gateway/example/ApprovalRequestControllers.kt
+++ b/example/src/main/kotlin/io/holixon/axon/gateway/example/ApprovalRequestControllers.kt
@@ -153,13 +153,17 @@ class ApprovalRequestReadController(
         ResponseTypes.instanceOf(ApprovalRequestQueryResult::class.java)
       )
       .thenApply { result ->
-        ok(
-          ApprovalRequestDto(
-            subject = result.payload.subject,
-            amount = result.payload.amount,
-            currency = result.payload.currency
+        if (result.payload != null) {
+          ok(
+            ApprovalRequestDto(
+              subject = result.payload.subject,
+              amount = result.payload.amount,
+              currency = result.payload.currency
+            )
           )
-        )
+        } else {
+          notFound().build()
+        }
       }
       .exceptionally { notFound().build() }
       .join()

--- a/example/src/main/kotlin/io/holixon/axon/gateway/example/Projection.kt
+++ b/example/src/main/kotlin/io/holixon/axon/gateway/example/Projection.kt
@@ -35,8 +35,7 @@ class ApprovalRequestProjection(
    */
   @QueryHandler
   fun getOneInResult(query: ApprovalRequestQuery): ApprovalRequestQueryResult {
-    require(storage.containsKey(query.requestId)) { "Approval request with id ${query.requestId} not found." }
-    return ApprovalRequestQueryResult(payload = storage.getValue(query.requestId), revisionValue = revisions.getValue(query.requestId))
+    return ApprovalRequestQueryResult(payload = storage[query.requestId], revisionValue = revisions[query.requestId] ?: RevisionValue.NO_REVISION)
   }
 
   /**
@@ -45,9 +44,8 @@ class ApprovalRequestProjection(
    * @return approval request.
    */
   @QueryHandler
-  fun getOne(query: ApprovalRequestQuery): QueryResponseMessage<ApprovalRequest> {
-    require(storage.containsKey(query.requestId)) { "Approval request with id ${query.requestId} not found." }
-    return QueryResponseMessageResponseType.asQueryResponseMessage(storage.getValue(query.requestId), revisions.getValue(query.requestId).toMetaData())
+  fun getOne(query: ApprovalRequestQuery): QueryResponseMessage<ApprovalRequest?> {
+    return QueryResponseMessageResponseType.asQueryResponseMessage(storage[query.requestId], (revisions[query.requestId] ?: RevisionValue.NO_REVISION).toMetaData())
   }
 
   /**
@@ -155,6 +153,6 @@ data class ApprovalRequestQuery(
  * @param revision of the projection.
  */
 data class ApprovalRequestQueryResult(
-  val payload: ApprovalRequest,
+  val payload: ApprovalRequest?,
   override val revisionValue: RevisionValue
 ) : Revisionable

--- a/example/src/test/kotlin/io/holixon/axon/gateway/example/JacksonApplicationITest.kt
+++ b/example/src/test/kotlin/io/holixon/axon/gateway/example/JacksonApplicationITest.kt
@@ -1,25 +1,36 @@
 package io.holixon.axon.gateway.example
 
+import org.awaitility.Awaitility
 import org.junit.jupiter.api.Disabled
 import org.junit.jupiter.api.Test
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.boot.test.context.SpringBootTest
 import org.springframework.test.annotation.DirtiesContext
 import org.springframework.test.context.ActiveProfiles
+import java.time.Duration
+import java.util.*
 import kotlin.test.assertNotNull
 
 @SpringBootTest
-@ActiveProfiles("jackson")
+@ActiveProfiles("jackson", "inmem")
 @DirtiesContext
-@Disabled // FIXME: #115
 internal class JacksonApplicationITest {
 
   @Autowired
   private lateinit var scenario: TestScenario
 
+  val requestId = UUID.randomUUID().toString()
+
   @Test
   fun `should create and query for request using jackson`() {
-    val requestId = scenario.createRequest()
-    assertNotNull(scenario.queryForRequest(requestId), "Could not find request for id $requestId")
+    // create request
+    Thread {
+      Thread.sleep(2_000)
+      scenario.createRequest(requestId)
+    }.start()
+
+    Awaitility.await("Could not find request for id $requestId").atMost(Duration.ofSeconds(13)).untilAsserted {
+      assertNotNull(scenario.queryForRequest(requestId))
+    }
   }
 }

--- a/example/src/test/kotlin/io/holixon/axon/gateway/example/TestScenario.kt
+++ b/example/src/test/kotlin/io/holixon/axon/gateway/example/TestScenario.kt
@@ -20,8 +20,7 @@ internal class TestScenario(
     const val revision = 1L
   }
 
-  fun createRequest(): String {
-    val requestId = UUID.randomUUID().toString()
+  fun createRequest(requestId: String = UUID.randomUUID().toString()): String {
     return commandGateway.send<String>(
       GenericCommandMessage.asCommandMessage<CreateApprovalRequestCommand>(
         CreateApprovalRequestCommand(
@@ -45,7 +44,14 @@ internal class TestScenario(
           .asCommandMessage<ApprovalRequestQuery>(ApprovalRequestQuery(requestId.trim()))
           .withMetaData(RevisionQueryParameters(revision).toMetaData()),
         QueryResponseMessageResponseType.queryResponseMessageResponseType<ApprovalRequest>()
-      )
+      ).handle { result, throwable ->
+        if (throwable == null) {
+          result
+        } else {
+          logger.error(throwable) { "Error occured during query." }
+          null
+        }
+      }
       .join()
   }
 }

--- a/example/src/test/kotlin/io/holixon/axon/gateway/example/XstreamApplicationITest.kt
+++ b/example/src/test/kotlin/io/holixon/axon/gateway/example/XstreamApplicationITest.kt
@@ -1,23 +1,37 @@
 package io.holixon.axon.gateway.example
 
+import org.awaitility.Awaitility
 import org.junit.jupiter.api.Disabled
 import org.junit.jupiter.api.Test
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.boot.test.context.SpringBootTest
 import org.springframework.test.annotation.DirtiesContext
+import org.springframework.test.context.ActiveProfiles
+import java.lang.Thread.sleep
+import java.time.Duration
+import java.util.*
 import kotlin.test.assertNotNull
 
 @SpringBootTest
 @DirtiesContext
-@Disabled // FIXME: #115
+@ActiveProfiles("inmem")
 internal class XstreamApplicationITest {
 
   @Autowired
   private lateinit var scenario: TestScenario
 
+  val requestId = UUID.randomUUID().toString()
+
   @Test
   fun `should create and query for request using XStream`() {
-    val requestId = scenario.createRequest()
-    assertNotNull(scenario.queryForRequest(requestId), "Could not find request for id $requestId")
+    // create request
+    Thread {
+      sleep(2_000)
+      scenario.createRequest(requestId)
+    }.start()
+
+    Awaitility.await("Could not find request for id $requestId").atMost(Duration.ofSeconds(13)).untilAsserted {
+      scenario.queryForRequest(requestId)
+    }
   }
 }

--- a/example/src/test/kotlin/io/holixon/axon/gateway/example/XstreamApplicationITest.kt
+++ b/example/src/test/kotlin/io/holixon/axon/gateway/example/XstreamApplicationITest.kt
@@ -31,7 +31,7 @@ internal class XstreamApplicationITest {
     }.start()
 
     Awaitility.await("Could not find request for id $requestId").atMost(Duration.ofSeconds(13)).untilAsserted {
-      scenario.queryForRequest(requestId)
+      assertNotNull(scenario.queryForRequest(requestId))
     }
   }
 }

--- a/example/src/test/resources/application.yml
+++ b/example/src/test/resources/application.yml
@@ -1,0 +1,4 @@
+axon-gateway:
+  query:
+    revision-aware:
+      default-query-timeout: 5_000

--- a/extension/core/src/main/kotlin/io/holixon/axon/gateway/query/QueryResponseMessageResponseType.kt
+++ b/extension/core/src/main/kotlin/io/holixon/axon/gateway/query/QueryResponseMessageResponseType.kt
@@ -37,7 +37,8 @@ class QueryResponseMessageResponseType<T : Any> : AbstractResponseType<T> {
      * @return message to be returned by the query handler.
      */
     @JvmStatic
-    inline fun <reified T : Any> asQueryResponseMessage(payload: Any?): QueryResponseMessage<T> = GenericQueryResponseMessage.asNullableResponseMessage(T::class.java, payload)
+    inline fun <reified T : Any?> asQueryResponseMessage(payload: Any?): QueryResponseMessage<T> = GenericQueryResponseMessage
+      .asNullableResponseMessage(T::class.java, payload)
 
     /**
      * Creates a query response message with given payload and metadata.
@@ -47,7 +48,7 @@ class QueryResponseMessageResponseType<T : Any> : AbstractResponseType<T> {
      * @return message to be returned by the query handler.
      */
     @JvmStatic
-    inline fun <reified T : Any> asQueryResponseMessage(payload: Any?, metaData: MetaData): QueryResponseMessage<T> = GenericQueryResponseMessage
+    inline fun <reified T : Any?> asQueryResponseMessage(payload: Any?, metaData: MetaData): QueryResponseMessage<T> = GenericQueryResponseMessage
         .asNullableResponseMessage(T::class.java, payload)
         .andMetaData(metaData)
 
@@ -58,8 +59,8 @@ class QueryResponseMessageResponseType<T : Any> : AbstractResponseType<T> {
      * @return message to be passed to query update emitter.
      */
     @JvmStatic
-    inline fun <reified T : Any> asSubscriptionUpdateMessage(payload: Any?): SubscriptionQueryUpdateMessage<T> = GenericSubscriptionQueryUpdateMessage
-        .asUpdateMessage<T>(payload)
+    inline fun <reified T : Any?> asSubscriptionUpdateMessage(payload: Any?): SubscriptionQueryUpdateMessage<T> = GenericSubscriptionQueryUpdateMessage
+        .asUpdateMessage(payload)
 
     /**
      * Creates a query subscription update message with given payload and metadata.

--- a/pom.xml
+++ b/pom.xml
@@ -19,9 +19,9 @@
   </modules>
 
   <properties>
-    <axon.version>4.5.10</axon.version>
+    <axon.version>4.5.14</axon.version>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-    <kotlin.version>1.6.21</kotlin.version>
+    <kotlin.version>1.7.10</kotlin.version>
     <kotlin-logging.version>2.1.23</kotlin-logging.version>
 
     <java.version>11</java.version>
@@ -34,7 +34,7 @@
     <log4j.version>2.13.3</log4j.version>
     <dokka.version>1.6.21</dokka.version>
 
-    <springboot.version>2.7.0</springboot.version>
+    <springboot.version>2.7.1</springboot.version>
 
     <!-- Skip KDoc generation by default -->
     <dokka.skip>true</dokka.skip>

